policy/release/tasks.rego

```diff
@@ -227,6 +227,78 @@ deny contains result if {
 	)
 }
 
+# METADATA
+# title: Task marked as deprecated
+# description: >-
+#   The Tekton Task used is marked as deprecated, upgrade to a newer version.
+# custom:
+#   short_name: deprecated
+#   failure_msg: >-
+#     Task %q is used by pipeline task %q is marked as deprecated and it will expire
+#     on %s.
+#   solution: >-
+#     Upgrade to a newer version of the Task.
+#   collections:
+#   - redhat
+#   depends_on:
+#   - tasks.pipeline_has_tasks
+#
+warn contains result if {
+	some att in lib.pipelinerun_attestations
+	some task in tkn.tasks(att)
+
+	annotations := tkn.task_annotations(task)
+
+	deprecated_on := annotations[_expires_on_annotation]
+	deprecated_on_rfc3339 := sprintf("%s%s", [
+		deprecated_on,
+		substring("0000-00-00T00:00:00Z", count(deprecated_on), 20 - count(deprecated_on)),
+	])
+
+	time.parse_rfc3339_ns(deprecated_on_rfc3339) > lib.time.effective_current_time_ns
+
+	result := lib.result_helper_with_term(
+		rego.metadata.chain(),
+		[tkn.task_name(task), tkn.pipeline_task_name(task), deprecated_on],
+		tkn.task_name(task),
+	)
+}
+
+# METADATA
+# title: Task is expired
+# description: >-
+#   The Tekton Task used is expired, upgrade to a newer version.
+# custom:
+#   short_name: expired
+#   failure_msg: Task %q is used by pipeline task %q has expired on %s.
+#   solution: >-
+#     Upgrade to a newer version of the Task.
+#   collections:
+#   - redhat
+#   depends_on:
+#   - tasks.pipeline_has_tasks
+#
+deny contains result if {
+	some att in lib.pipelinerun_attestations
+	some task in tkn.tasks(att)
+
+	annotations := tkn.task_annotations(task)
+
+	deprecated_on := annotations[_expires_on_annotation]
+	deprecated_on_rfc3339 := sprintf("%s%s", [
+		deprecated_on,
+		substring("0000-00-00T00:00:00Z", count(deprecated_on), 20 - count(deprecated_on)),
+	])
+
+	time.parse_rfc3339_ns(deprecated_on_rfc3339) <= lib.time.effective_current_time_ns
```

Comment:
Let's avoid doing any sort of date processing in rego.

I think here we can combine both the warn and deny policy rules into a single deny rule. Then, simply set the `effective_on` in the result based on the value from the deprecated label.

Reply:
+1